pkg/virt-api/rest/subresource.go

```diff
@@ -314,7 +314,8 @@ func (app *SubresourceAPIApp) StartVMRequestHandler(request *restful.Request, re
 				return
 			}
 		} else {
-			if vmi.IsFinal() {
+			if (runStrategy == v1.RunStrategyRerunOnFailure && vmi.Status.Phase == v1.Succeeded) ||
```

Comment:
That looks better now. Let's now be explicit by returning an error if you try to do a `start` operation on a VMI which is in failed state and has `RunStrategyRerunOnFailure`. Return something like `Only VMIs in succeeded state support a start operation with the 'RerunOnFailure' run policy.` We really need to be explicit on imperative actions, if they have an effect or not.

Once this is there, it looks good to me.

Reply:
what about when ```!vmi.IsFinal() && runStrategy == v1.RunStrategyManual```

it looks like we're we're sending a start request in that scenario, when it should arguably be an error if we're being explicit about imperative actions

Reply:
I think we're starting to go down a dark path here. On the virt-controller side, if a start request is issued and the VMI is present, then the start request has been met and gets cleared. This of course also applies to the `vmi.IsFinal() && runStrategy == v1.RunStrategyManual` case. But in that case something unexpected happens--because the user asked for the VMI to be started when the next thing that will happen is a deletion.

What I want to avoid is a scenario where virt-api refuses to issue a start request (due to a bug?) where it should have. In such a case, there'd be no workaround.

I think for any `!vmi.IsFinal()` case, we should pass the request over to virt-controller. Let's keep as much of the opinionated state logic there as we possibly can.

Reply:
I think there's a miscommunication. Take a look at @rmohr's comment again. This part specifically 

> We really need to be explicit on imperative actions, if they have an effect or not.

What @rmohr pointed out here exposed that we're behaving inconsistently with respects to how we report errors back to the user for imperative actions. You fixed his initial request there, but I'm unsure the intent behind his comment is being fully observed throughout the Start/Stop/Restart subresource endpoints.

Basically, should we be propagating errors back to the user when the they request to perform an imperative action that results in the VM remaining in its current state (starting a started vm, or stopping a stopped vm)? Right now we aren't consistent in how this behaves.

Reply:
This isn't an error condition though.

I've just double checked and I say we are consistent. the errors we report right now are for clearly nonsensical things like "start request for an always-running runstrategy" or "stop request for the halted runstrategy"...

Remember, RunStrategy does not indicate the running state, only the intent. We only report back errors when somebody issues a nonsensical *intent*. Let's really not get into the game of virt-api arbitrating what requests its willing to fill when virt-controller can sort things out against the current state of the cluster. Simpler is more robust

Reply:
> I've just double checked and I say we are consistent.

for manual mode... we report an error if we attempt to stop a VMI that's already stopped... we don't report an error if we attempt to start a VMI that's already started.

how is that consistent? it has nothing to do with the error message, it's about when errors are returned or not. I don't know how I can be more clear in my explanation

Reply:
@davidvossel is right. That should be improved. @stu-gott can you create an issue to follow up on this?